### DISCUSSION TITLE
Discussion about implementing distributed query execution with DataFusion

### DISCUSSION THREAD

**Discussion by alamb** [apache/datafusion#8139] [General]
I broke this discussion out of https://github.com/apache/arrow-datafusion/issues/6782#issuecomment-1805487141 as there is already a lot of discussion on that ticket and I thought putting this thread in a new discussion would make it easier to collaborate

@waynexia  writes:

> Hi, I want to share a use case about §7.6 Query Rewrites / Optimizer Passes.
> 
> Besides making domain-specific features, the optimizer interface is also an excellent place to make the plan "scale" across different compute nodes. I've been exploring how to run distributed queries over DataFusion. It is not something similar to Ballista, but an optimizer to "fuse" several engine instances across compute nodes ([RFC](https://github.com/GreptimeTeam/greptimedb/blob/develop/docs/rfcs/2023-05-09-distributed-planner.md)). At this stage it can push down some plans to the node which is closer to data, and then collect and merge the final result using the remaining part of plan. It's also implemented as an optimizer (an analyzer rule actually).

**Comment by alamb** [comment]
This is a very neat idea @waynexia  -- indeed we have similar rewrite passes in IOx (that operate on `ExecutionPlan`s (not `LogicalPlan`s) that implement various pushdown / pull op optimizations. 

It almost sounds to me like the ([RFC](https://github.com/GreptimeTeam/greptimedb/blob/develop/docs/rfcs/2023-05-09-distributed-planner.md)) is implementing a "pull up" in that it is pulling the "get data from remote node" **up** in the plan (and thus pushing parts of the plan's computation **down** to the remote nodes.

It might make sense to consider doing this optimization on the ExecutionPlan if you find you don't have enough information  (like partitioning) to push down operations to the execution nodes

**Comment by waynexia** [comment]
>indeed we have similar rewrite passes in IOx (that operate on `ExecutionPlan`s (not `LogicalPlan`s) that implement various pushdown / pull op optimizations.

That's great 👍 I also find some similar logic in the existing codebase, the multi-phase aggregation [you mentioned](https://github.com/apache/arrow-datafusion/issues/7960#issuecomment-1784081531). The overall idea is the same: transform an operation into multiple steps where some can be executed in parallel.

> It almost sounds to me like the ([RFC](https://github.com/GreptimeTeam/greptimedb/blob/develop/docs/rfcs/2023-05-09-distributed-planner.md)) is implementing a "pull up" in that it is pulling the "get data from remote node" **up** in the plan (and thus pushing parts of the plan's computation **down** to the remote nodes.

Cool, that's right. I call it "push down" because it's easier to understand. The RFC and implementation are pulling the "merge node" up in actual. The "commutativity" part in RFC is defined on the "merge node" with all other plans. With it, we can know how to or if it is possible to exchange the "merge node" with its parent node. The more we pull the "merge node" up the later we have to collect and aggregate the results.
 
> It might make sense to consider doing this optimization on the ExecutionPlan if you find you don't have enough information (like partitioning) to push down operations to the execution nodes

I go with logical plan with two points: (1) it's more generic, and doesn't contain some execution information (hence the rule might be more straightforward). (2) it can be processed earlier than the physical plan.

In the logical planning phase, we might lack physical information like the partitioning you said. But others like how the data is distributed across nodes, how many nodes we will use are available. Those "logical" information might be enough for us to play with logical plan. And then we can leave physical planning and optimization to the node that will execute the sub-plan, where it has more execution information.

One big part of this is we have to define a good and complete rule set of commutativity. I've made some but it's still very scarce. And another part is substrait, the proto to send sub-plan. I should admit it's not a good way compared with plain protobuf at the current stage, so many plan/exprs are absent and brings many limitations.

**Comment by andygrove** [comment]
Loosely related to this, I propose that we create a `datafusion-distributed` crate with utilities to help people build distributed systems with DataFusion, so things like the distributed query planning that is being discussed here, and shuffle mechanisms. I have an initial PR at https://github.com/apache/datafusion/pull/11070
